### DISCUSSION TITLE
Perform a check before playing a video in a playlist

### DISCUSSION THREAD

**Discussion by mrjackcitrus** [videojs/video.js#7544] [General]
I have a videoJS playlist loaded in the video player.

Each video source in the playlist consists of a presigned URL that expires in X number of seconds.

I know exactly when the videos expire and I can obtain fresh signatures to extend their expiry time if required. There will be scenarios when this needs to happen, for example, if the viewer pauses the playlist and comes back to it later.

Upon the user pressing play (or selecting a new item on the playlist) I want to perform something like the following pseudocode:

```
if(sourcesHaveExpired()){

    let updatedPlaylist = rebuildPlaylistWithNewPresignedUrls();
    
    window.videoPlayer.playlist(updatedPlaylist);

    // continue playing here, somehow

}
```

I tried listening out for videoJS's `play` event in the hope I could perform my expiry check here and then somehow stop the rest of the default chain of events from occurring if they had expired. But so far I've not been able to find a way to stop this.

Perhaps I need to reconfigure the event listeners for the player's default controls?

Any advice would be gratefully welcomed.

Many thanks

**Comment by gkatsev** [comment]
You probably want to look into middleware and either use the `setSource` piece to update the actual url or use the `callPlay` piece to intercept the play method call.
See https://docs.videojs.com/tutorial-middleware.html for more info.

**Reply by mrjackcitrus** [reply]
Thanks, that gives me something to go on. If I find a solution I'll post it here for posterity.

**Comment by mrjackcitrus** [comment]
Thanks for pointing me in the right direction, @gkatsev.

As per your recommendation, I have used middleware to achieve what I was trying to do.

Here's what I did for anyone that might be interested in the future (it's pseudocode, but you'll get the idea).

```
// Define middleware

videojs.use('*', {
    callPlay: function() {

        if( hasMyTokenExpired() ) {

            getNewTokenFromApi()
                .then( (token) => appendTokenToVideoSources(token) )
                .then( () => window.videoPlayer.play() );

            return videojs.middleware.TERMINATOR;

        }

    }
});

// Initialise video player

let videoPlayer = videojs( '#video-element-id');

```
